### PR TITLE
elvish 0.3 (new formula)

### PR DESCRIPTION
**Pull request by gaoyifan**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [X] Does your submission pass `brew audit --new-formula <formula>` (after doing `brew install <formula>`)?

---

### CODE REVIEW

File: Formula/elvish.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+require "language/go"
+
+class Elvish < Formula
+  desc "novel Unix shell"
```
Comment (zmwangx) [inline Formula/elvish.rb]: Use sentence case, i.e., capitalize "novel".

File: Formula/elvish.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  homepage "https://github.com/elves/elvish"
+  url "https://github.com/elves/elvish/archive/0.2.tar.gz"
```
Comment (zmwangx) [inline Formula/elvish.rb]: The current version is 0.3.

File: Formula/elvish.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  sha256 "00c0396eeb1ae42178e3c3a7c84068fe5b6b9ae866480ede4d7f00b379e32071"
+
```
Comment (zmwangx) [inline Formula/elvish.rb]: Remove this blank line.

File: Formula/elvish.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  head "https://github.com/elves/elvish.git"
+
+  depends_on "go" => :build
+
+  go_resource "github.com/elves/getopt" do
+    url "https://github.com/elves/getopt.git",
+      :reversion => "f91a7bf920995832d55a1182f26657bc975b9c24"
```
Comment (zmwangx) [inline Formula/elvish.rb]: Align `:` to the opening quote on the previous line. Same below.

File: Formula/elvish.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    mkdir_p buildpath/"src/github.com/elves/"
+    ln_sf buildpath, buildpath/"src/github.com/elves/elvish"
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    cd "src/github.com/elves/elvish" do
+      system "go", "build", "-o", "build/elvish"
```
Comment (zmwangx) [inline Formula/elvish.rb]:
```ruby
system "go", "build", "-o", bin/"elvish"
```

and remove the `bin.install` below.

File: Formula/elvish.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  go_resource "github.com/mattn/go-sqlite3" do
+    url "https://github.com/mattn/go-sqlite3.git",
+      :reversion => "3fb7a0e792edd47bf0cf1e919dfc14e2be412e15"
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    mkdir_p buildpath/"src/github.com/elves/"
```
Comment (zmwangx) [inline Formula/elvish.rb]:
```ruby
(buildpath/"src/github.com/elves").mkpath
```

is preferred.

**Comment by zmwangx** [comment]
A few comments.

Other than that, I would probably forgo the `cd` block and instead use

```ruby
system "make", "-C", "src/github.com/elves/elvish", "stub"
```

instead. But other maintainers might disagree with me here.

File: Formula/elvish.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    (buildpath/"src/github.com/elves/elvish").install Dir["*"]
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    system "go", "build", "-o", bin/"elvish", "github.com/elves/elvish"
+    system "make", "-C", "src/github.com/elves/elvish", "stub"
+    bin.install buildpath/"bin/elvish-stub"
```
Comment (zmwangx) [inline Formula/elvish.rb]: No longer need `buildpath/` here.

File: Formula/elvish.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  homepage "https://github.com/elves/elvish"
+  url "https://github.com/elves/elvish/archive/0.3.tar.gz"
+  sha256 "83d463667055f38367302921e8d87c274ae71fc777f6ffb176dad96e875378e8"
+  head "https://github.com/elves/elvish.git"
+
+  depends_on "go" => :build
+
+  go_resource "github.com/elves/getopt" do
+    url "https://github.com/elves/getopt.git",
+    :reversion => "f91a7bf920995832d55a1182f26657bc975b9c24"
```
Comment (ilovezfs) [inline Formula/elvish.rb]: reversion to the mean?

Comment (zmwangx) [inline Formula/elvish.rb]: Slipped right past my eyes... And indentation is still wrong.

```ruby
    url "https://github.com/elves/getopt.git",
        :reversion => "f91a7bf920995832d55a1182f26657bc975b9c24"
```

File: Formula/elvish.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  go_resource "github.com/mattn/go-sqlite3" do
+    url "https://github.com/mattn/go-sqlite3.git",
+    :reversion => "3fb7a0e792edd47bf0cf1e919dfc14e2be412e15"
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    (buildpath/"src/github.com/elves/elvish").install Dir["*"]
```
Comment (zmwangx) [inline Formula/elvish.rb]: Please use what you had before, with my proposed style change:

```ruby
    (buildpath/"src/github.com/elves").mkpath
    ln_sf buildpath, buildpath/"src/github.com/elves/elvish"
```

File: Formula/elvish.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    Language::Go.stage_deps resources, buildpath/"src"
+
```
Comment (zmwangx) [inline Formula/elvish.rb]: I'd rather see this blank line removed.

**Comment by zmwangx** [comment]
More comments. Please try not to make unnecessary changes where we don't ask you to change...

File: Formula/elvish.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    system "go", "build", "-o", bin/"elvish", "github.com/elves/elvish"
```
Comment (zmwangx) [inline Formula/elvish.rb]:
```ruby
system "go", "build", "-o", bin/"elvish"
```

should be enough. Apparently `go build` is smart enough to compile that directory.

File: Formula/elvish.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  homepage "https://github.com/elves/elvish"
+  url "https://github.com/elves/elvish/archive/0.3.tar.gz"
+  sha256 "83d463667055f38367302921e8d87c274ae71fc777f6ffb176dad96e875378e8"
+  head "https://github.com/elves/elvish.git"
+
+  depends_on "go" => :build
+
+  go_resource "github.com/elves/getopt" do
+    url "https://github.com/elves/getopt.git",
+        :reversion => "f91a7bf920995832d55a1182f26657bc975b9c24"
```
Comment (zmwangx) [inline Formula/elvish.rb]: `:revision`.

File: Formula/vaulted.rb [hunk 13]
```diff
@@ -63,7 +63,7 @@ def install
     ln_s buildpath, "src/github.com/miquella/vaulted"
     Language::Go.stage_deps resources, buildpath/"src"
 
-    system "go", "build", "-o", bin/"vaulted", "github.com/miquella/vaulted"
+    system "go", "build", "-o", bin/"vaulted"
```
Comment (zmwangx) [inline Formula/vaulted.rb]: Thanks, but please keep your new formula PR localized to your new formula...

File: Formula/elvish.rb [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    url "https://github.com/mattn/go-sqlite3.git",
+        :reversion => "3fb7a0e792edd47bf0cf1e919dfc14e2be412e15"
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    (buildpath/"src/github.com/elves").mkpath
+    ln_sf buildpath, buildpath/"src/github.com/elves/elvish"
+    Language::Go.stage_deps resources, buildpath/"src"
+    system "go", "build", "-o", bin/"elvish", "github.com/elves/elvish"
```
Comment (zmwangx) [inline Formula/elvish.rb]: What you really need to address is this line.

File: Formula/vaulted.rb (hunk 13 above)
Comment (gaoyifan) [inline Formula/vaulted.rb]: It's a mistake....

**Comment by zmwangx** [comment]
LGTM now. @ilovezfs More thoughts?

File: Formula/elvish.rb [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+        :revision => "3fb7a0e792edd47bf0cf1e919dfc14e2be412e15"
+  end
+
+  def install
+    ENV["GOPATH"] = buildpath
+    (buildpath/"src/github.com/elves").mkpath
+    ln_sf buildpath, buildpath/"src/github.com/elves/elvish"
+    Language::Go.stage_deps resources, buildpath/"src"
+    system "go", "build", "-o", bin/"elvish"
+    system "make", "-C", "src/github.com/elves/elvish", "stub"
```
Comment (ilovezfs) [inline Formula/elvish.rb]: this looks like it's doing a `go get` which is a no-no.

Comment (ilovezfs) [inline Formula/elvish.rb]: ah I see it's only the stub target.

File: Formula/elvish.rb [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    bin.install "bin/elvish-stub"
+  end
+
+  test do
+    assert_match "test message", shell_output("#{bin}/elvish -c 'echo test message'")
```
Comment (ilovezfs) [inline Formula/elvish.rb]: too wide

**Comment by ilovezfs** [comment]
Since this is a new formula, I'd prefer to see it using some form of go vendoring ...

File: Formula/elvish.rb (hunk 16 above)
Comment (zmwangx) [inline Formula/elvish.rb]: Just need to shorten the message. Like

```ruby
assert_match "hello", shell_output("#{bin}/elvish -c 'echo hello'")
```

File: Formula/elvish.rb [hunk 17]
```diff
@@ -0,0 +1,35 @@
+require "language/go"
+
+class Elvish < Formula
+  desc "Novel Unix shell"
```
Comment (ilovezfs) [inline Formula/elvish.rb]: maybe a few more words here?

**Comment by zmwangx** [comment]
@ilovezfs You mean you're willing to merge until you see go vendoring? Our policy wasn't like that two days ago...

**Comment by zmwangx** [comment]
I'll file an issue about go vendoring upstream, so let's move on.

**Comment by zmwangx** [comment]
Merged as 29ef89e. Thank you for your contribution to Homebrew!

**Comment by zmwangx** [comment]
Submitted upstream issue https://github.com/elves/elvish/issues/209.

**Pull request closed without merging**